kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,591 @@
+}
+
+local function insert_last_type(cache, name, qtype)
+    if valid_types[qtype] then
+        cache:set("last:" .. name, { ttl = 0 }, qtype)
+    end
+end
+
+
+local function get_last_type(cache, name)
+    return cache:get("last:" .. name)
+end
+
+
+-- insert hosts into cache
+local function init_hosts(cache, path, preferred_ip_type)
+    local hosts, err = utils.parse_hosts(path)
+    if not hosts then
+        ngx.log(ngx.WARN, "Invalid hosts file: ", err)
+        hosts = {}
+    end
+
+    if not hosts.localhost then
+        hosts.localhost = {
+          ipv4 = "127.0.0.1",
+          ipv6 = "[::1]",
+        }
+    end
+
+    local function insert_answer(name, qtype, address)
+        if not address then
+            return
+        end
+
+        local key = name .. ":" .. qtype
+        local answers = {{
+            name = name,
+            type = qtype,
+            address = address,
+            class = 1,
+            ttl = 0,
+        }}
+        cache:set(key, { ttl = 0 }, answers)
+    end
+
+    for name, address in pairs(hosts) do
+        name = name:lower()
+        if address.ipv4 then
+            insert_answer(name, TYPE_A, address.ipv4)
+            insert_last_type(cache, name, TYPE_A)
+        end
+        if address.ipv6 then
+            insert_answer(name, TYPE_AAAA, address.ipv6)
+            if not address.ipv4 or preferred_ip_type == TYPE_AAAA then
+                insert_last_type(cache, name, TYPE_AAAA)
+            end
+        end
+    end
+
+    return hosts
+end
+
+
+function _M.new(opts)
+    if not opts then
+        return nil, "no options table specified"
+    end
+
+    -- parse resolv.conf
+    local resolv, err = utils.parse_resolv_conf(opts.resolv_conf, enable_ipv6)
+    if not resolv then
+        ngx.log(ngx.WARN, "Invalid resolv.conf: ", err)
+        resolv = { options = {} }
+    end
+
+	-- init the resolver options for lua-resty-dns
+    local nameservers = opts.nameservers or resolv.nameservers
+    if not nameservers or #nameservers == 0 then
+        ngx.log(ngx.WARN, "Invalid configuration, no nameservers specified")
+    end
+
+    local r_opts = {
+        nameservers = nameservers,
+        retrans = opts.retrans or resolv.options.attempts or 5,
+        timeout = opts.timeout or resolv.options.timeout or 2000,   -- ms
+        no_random = opts.no_random or not resolv.options.rotate,
+    }
+
+    -- init the mlcache
+    local lock_timeout = r_opts.timeout / 1000 * r_opts.retrans + 1 -- s
+
+    local cache, err = mlcache.new("dns_cache", "kong_dns_cache", {
+        lru_size = opts.cache_size or 10000,
+        ipc_shm = "kong_dns_cache_ipc",
+        resty_lock_opts = {
+            timeout = lock_timeout,
+            exptimeout = lock_timeout + 1,
+        },
+    })
+    if not cache then
+        return nil, "could not create mlcache: " .. err
+    end
+
+    cache:purge(true)
+
+    -- TODO: add an async task to call cache:update() to update L1/LRU-cache
```

Comment:
There should be no need for calling `mlcache:update()` when mlcache is properly used. Calling `:update()` should only be necessary when using `:set()`, `:delete()`, or `:purge()` when the L1 cache (worker Lua VM) wasn't empty. Unless the above mlcache instantiation will be made in an already-running worker, then the `:purge()` should not be necessary, as the worker should not have any value in L1 anyway.

Note that calling `:purge()` also clears the L2 cache. This means that a crashed worker that gets restarted by the master process will clear the L2 cache when instantiating the new client, thus clearing the L2 cache of *all other workers*. This probably should require more consideration.

mlcache is a very complete library but using it properly requires *deep* study and understanding of its design pattern. I strongly advise a thorough read of the documentation and codebase.

Reply:
Thanks for the advice, I will carefully consider the need for calling `mlcache:update()` and prioritize avoiding the use of `:update()`, striving to meet it through modifications in the business logic.

Reply:
> There should be no need for calling mlcache:update() when mlcache is properly used. Calling :update() should only be necessary when using :set(), :delete(), or :purge() when the L1 cache (worker Lua VM) wasn't empty. Unless the above mlcache instantiation will be made in an already-running worker, then the :purge() should not be necessary, as the worker should not have any value in L1 anyway.

See this comment: https://github.com/Kong/kong/pull/12305#discussion_r1505328799. I recently spoke with @ADD-SP. Since it is still necessary to support the use of and asynchronously update stale records, we have to make additional `mlcache:set()` calls to update the asynchronously updated records instead of the already cached stale records.